docs/guides/extending-beluga.md

```diff
@@ -1,3 +1,264 @@
+
+### Implementation
+
+In this library, motion models are implemented as callable objects that take a control action and return another callable, a state sampling function.
+This function can be applied to each particle, generating a random sample given the particle's previous state and a random number generator.
+
+Before diving into the implementation, we should see how the motion model is used in the particle filter algorithm shown above:
+
+```cpp
+particles |= beluga::actions::propagate(motion_model(control));
+```
+
+This functional composition can be explained as follows:
+
+1. The motion model uses the control value as input to generate a function that dictates how each particle's state should be adjusted.
+2. The `propagate` function takes the state sampling function and creates a range action that can be applied to a range of particles.
+3. The `|=` applies the range action to the entire set of particles, updating each one according to the model.
+
+This is the basic template we will use for creating a custom motion model:
+
+```cpp
+struct MotionModel {
+ public:
+  using control_type = ...;
+  using state_type = ...;
+
+  auto operator()(const control_type& action) const {
+    ...
+    return [=](state_type state, auto& engine) -> state_type { ... };
+  }
+};
+```
+
+Let's now implement a simple motion model of an agent moving across a 1-dimensional world. For this we will use the following definitions:
+
+- The state will be the position of the agent along an infinite line. Represented as a `double`.
+- The control will be the desired displacement of the agent in that line. Represented as a `double`.
+- The uncertainty associated with the agent reaching its intended position after a control command is issued is modeled using a Gaussian distribution.
+- The spread of the Gaussian distribution is controlled by a standard deviation, which is a constant parameter that can be specified when initializing the model.
+
+The resulting implementation could look like this:
+
+```cpp
+struct MotionModel {
+ public:
+  using control_type = double;
+  using state_type = double;
+
+  explicit SimpleMotionModel(double stddev) : stddev_{stddev} {}
+
+  auto operator()(control_type action) const {
+    auto param = typename std::normal_distribution<double>::param_type{action, stddev_};
+
+    return [param, action](state_type position, auto& engine) -> state_type {
+      static thread_local auto distribution = std::normal_distribution<double>{};
+      return position + distribution(engine, param);
+    };
+  }
+
+ private:
+  double stddev_;
+};
+```
+
+The implementation of the `MotionModel` demonstrates a straightforward yet powerful way to integrate custom motion dynamics into the Beluga library. This library already provides implementations that can be used as reference or starting point for custom models. See for example:
+
+- `beluga::DifferentialDriveModel`: Sampled odometry model for a differential drive robot.
```

Comment:
FWIW these should be links, but I have to fix [`doxylink`](https://github.com/sphinx-contrib/doxylink) first. It chokes on Beluga code right now, it is too complex for `doxylink`'s implementation :roll_eyes: